### DISCUSSION TITLE
Questions regarding this project

### DISCUSSION THREAD

**Discussion by Raj2032** [slint-ui/slint#839] [General] [answered]
Hey guys I have some questions regarding this project:

1. Is it retained, or immediate or a combination of both (like [`iced`](https://github.com/iced-rs/iced) GUI library does?)
2. Is it backend agnostic like [`egui`](https://github.com/emilk/egui) is?
3. Which library has a more performance overhead if animations is involved, `sixyfps` or [`egui`](https://github.com/emilk/egui)? Can I achieve full very smooth animations like in `egui`'s [demo](https://emilk.github.io/egui/#demo) and can I make it look similar to this?
4. Am I able to really design exactly how I want my UI to be designed, for example like this:

![image](https://user-images.githubusercontent.com/29293830/149652841-0db2c23b-b56d-46af-8438-aa066bb31035.png)

?

5. Does it lock you into some kind of framework, for example like how [`iced`](https://github.com/iced-rs/iced) does or the [`bevy`](https://bevyengine.org/) game engine does it with their ECS model or can I code more freely like [`fltk-rs`](https://github.com/fltk-rs/fltk-rs) (here is their code sample)

```rust
use fltk::{app, prelude::*, window::Window};

fn main() {
    let app = app::App::default();
    let mut wind = Window::new(100, 100, 400, 300, "Hello from rust");
    wind.end();
    wind.show();
    app.run().unwrap();
}
```

**Comment by tronical** [comment, marked as answer]
> Hey guys I have some questions regarding this project:
> 
> 1. Is it retained, or immediate or a combination of both (like [`iced`](https://github.com/iced-rs/iced) GUI library does?)

It is designed to be retained mode.

> 2. Is it backend agnostic like [`egui`](https://github.com/emilk/egui) is?

Yes, have two two backends and it is generally designed to allow for multiple.

> 3. Which library has a more performance overhead if animations is involved, `sixyfps` or [`egui`](https://github.com/emilk/egui)?

That is hard to say and depends on the complexity of our scene and your compute environment.

> Can I achieve full very smooth animations like in `egui`'s [demo](https://emilk.github.io/egui/#demo) and can I make it look similar to this?

I believe so, yes.

> 4. Am I able to really design exactly how I want my UI to be designed, for example like this:
> 
> ![image](https://user-images.githubusercontent.com/29293830/149652841-0db2c23b-b56d-46af-8438-aa066bb31035.png)
> 
> ?

Yes, that is our objective. We may not have the features set of egui just yet, but we're working towards it while maintaining our design oriented approach (as opposed to programming the user).

> 5. Does it lock you into some kind of framework, for example like how [`iced`](https://github.com/iced-rs/iced) does or the [`bevy`](https://bevyengine.org/) game engine does it with their ECS model or can I code more freely like [`fltk-rs`](https://github.com/fltk-rs/fltk-rs) (here is their code sample)
> 
> ```rust
> use fltk::{app, prelude::*, window::Window};
> 
> fn main() {
>     let app = app::App::default();
>     let mut wind = Window::new(100, 100, 400, 300, "Hello from rust");
>     wind.end();
>     wind.show();
>     app.run().unwrap();
> }
> ```

Well, we have not designed our API to be ... API agnostic :-). So we do require the programmer to use certain methods or APIs, partly because of our design choice and partly because of the underlying libraries that we use. For example at the moment we require running the event loop in the main thread. We do however provide functions to allow for communication between other threads, to make an integration as easy as possible. For example cargo-ui uses that to do its work using using async functions driven by tokio, in a separate thread, and it communicates with the SixtyFPS UI.

That said, you're writing `main()`, and you decide when to call functions. Just eventually you'll have to call `sixtyfps::run_event_loop()` and then you'll receive callbacks that you may have previously installed on the UI. Take a look at [Cargo UI](https://github.com/sixtyfpsui/cargo-ui) if you'd like to see some more complex code :-)

**Reply by Raj2032** [reply]
Thanks mate  for your help =D
